Add precision and recall at k example in the FAQ

mahermalaeb opened the pull request:
This issue resolves the discussion in #70.

For cases where denominator is zero, the case is handled as explained [here](https://stats.stackexchange.com/questions/1773/what-are-correct-values-for-precision-and-recall-in-edge-cases). I believe this is different from how it is handled in the lyst/lightfm library [here](https://github.com/lyst/lightfm/blob/master/lightfm/evaluation.py)

NicolasHug commented:
Hi Maher, 
Thanks for this!

I have a few remarks:
- I'm not sure using numpy is really necessary here. Numpy might be efficient but it really obfuscates the code and in this case there is no need to use deeply optimized code. Plus, as this would be an FAQ entry, I would rather sacrifice efficiency for simplicity (and "understandability"). I really think we could implement the two metrics by adding a few lines of code the `top_n` example.
- Your description and implementation of Recall@k and Precision@k do not fit my understanding of these metrics: you use the `relevant@k` set (see [here](https://github.com/NicolasHug/Surprise/pull/74/files#diff-4942885287e043283c7cc7a43f93c3a3R66)), but  to me, this should just be the set of `relevant` items (not `relevant@k`), i.e. the set of items such that `true_r >= threshold`. For you this is the set of items such that  `true_r >= threshold AND rank <= k`, where the rank is decided by the prediction value. I may be wrong but I understand that [your resource](https://grouplens.org/site-content/uploads/evaluating-TOIS-20041.pdf) (and other stuff that I have read)  simply uses `relevant`.
- I think returning the two `precisions` and `recalls` dictionaries would be enough rather than returning the average for all users. Then the user could compute on its own the mean, the standard deviation, etc.

Thanks!

mahermalaeb commented:
Hello Nicolas,

Thank you for your thorough review. I have indeed misunderstood some of the readings I did on this topic where I ended up with an over complicated algorithm. The algorithm should be now correct and simple enough. 
* I have added only few lines of code to the top-N recommendation example
* I am no longer using numpy
* The algorithm should now be inline with your description. To be sure I have validated its results with the results presented in a [presentation from UNC](https://ils.unc.edu/courses/2013_spring/inls509_001/lectures/10-EvaluationMetrics.pdf) on slide 25. You can use the below code example to reproduce the results
```
def pred(true_r, est, u0=None):
    """Just a small tool to build a prediction with appropriate format."""
    return (u0, None, true_r, est, None)

predictions = [pred(4, 4.99, u0='u1'), pred(3, 4.98, u0='u1'), pred(4, 4.97, u0='u1'), pred(4, 4.96,
               u0='u1'), pred(4, 4.95, u0='u1'), pred(4, 4.94, u0='u1'), pred(4, 4.93, u0='u1'), pred(3, 4.92,
               u0='u1'), pred(4, 4.91, u0='u1'), pred(3, 4.90, u0='u1'), pred(4, 4.89, u0='u1'), pred(3, 4.88,
               u0='u1'), pred(3, 4.87, u0='u1'), pred(4, 4.86, u0='u1'), pred(3, 4.85, u0='u1'), pred(3, 4.84,
               u0='u1'), pred(3, 4.83, u0='u1'), pred(3, 4.82, u0='u1'), pred(3, 4.81, u0='u1'), pred(4, 4.80,
               u0='u1'), pred(4, 4.79, u0='u1'), pred(4, 4.78, u0='u1'), pred(4, 4.77, u0='u1'), pred(4, 4.76,
               u0='u1'), pred(4, 4.75, u0='u1'), pred(4, 4.74, u0='u1'), pred(4, 4.73, u0='u1'), pred(4, 4.72,
               u0='u1'), pred(4, 4.71, u0='u1'), pred(4, 4.70, u0='u1')]
for i in range(1,11):
    print(precision_recall_at_k(predictions, k=i, threshold=4))
```
* The function now returns a dictionary of precision@k, recall@k tuple. It is no longer calculating the mean as you suggested.
* Travis checks has passed. Weird issue though is that pep8 tests are green locally but fails on Travis. 

Sorry for any inconveniences I caused in the my first attempt. Let's see the second!

Regards,
Maher

NicolasHug commented:
Hi Maher,
Thanks for these changes, the code is much easier to read now!
I still have a few remarks:
- Converting to float before a division isn't necessary as you have imported `division` from the `__future__` module: this ensures that Python2 has the same division scheme as Python3.
- Counting elements in a list satisfying a given property can be done in a more Pythonic fashion with 
`sum(condition(element) for elemnt in my_list)`. E.g. [lines 36 and 37](https://github.com/NicolasHug/Surprise/pull/74/files#diff-4942885287e043283c7cc7a43f93c3a3R36) could be merged into a one-liner: `n_rel = sum((true_r >= threshold) for (_, _, true_r) in user_ratings)`
- There is no need to store `iid` in the `user_ratings` list as we don't use it
- `top_n` may be a confusing name for the first dict (we're not retrieving top_n items, after all). Also, some of the comments in the code come from the top-n example and are not relevant here
- If I'm not mistaken, you are computing precision@k and recall@k this way:
  - precision@k = relevant@k / k
  - recall@k = relevant@k / relevant
This still does not fit what is explained in the resource you showed me. The relevant@k set should never be used. Instead, I understand that precision@k and recall@k are defined as follows:
  - precision@k = (relevant AND recommended@k) / recommended@k
  - recall@k = (relevant AND recommended@k) / relevant
where `relevant` corresponds to what you compute on [lines 36-37](https://github.com/NicolasHug/Surprise/pull/74/files#diff-4942885287e043283c7cc7a43f93c3a3R36), i.e. the set of items with`true_r >= threshold`, and recommended@k is the set of items such that `est >= threshold AND rank <= k`, where the rank is decided by the prediction value. It could be computed using e.g. `recommended = sum(est >= threshold for (est, _) in user_ratings[:k])`

mahermalaeb commented:
Hello Nicolas,

* float division is now fixed.
* Counting is now more pythonic as you suggested
* unused variables are now removed and comments reviewed.
* top-n variable is renamed
* Concerning the algorithm logic:
  * My previous implementation assumed that the top-k items are all recommended. This will make `recommended@k = k` and `relevant and recommended @k = relevant @k`.
  * As you suggested I have relaxed this condition and I am now using only the recommended @k items from the top-k recommendations thus the below should now be satisfied
    * precision@k = (relevant AND recommended@k) / recommended@k
    * recall@k = (relevant AND recommended@k) / relevant

I hope this time we are almost done :). Thanks for the advises and comments as I am learning a lot from them.

Regards,
Maher

NicolasHug commented on doc/source/FAQ.rst:
```diff
@@ -20,6 +20,23 @@ top-10 prediction for each user.
 
+.. _precision_recall_at_k:
+
+How to evaluate an algorithm using precision and recall at k metrics
```
How to compute precision@k and recall@k

mahermalaeb replied:
Fixed

NicolasHug commented on examples/precision_recall_at_k.py:
```diff
@@ -0,0 +1,73 @@
+from surprise import SVD
+
+
+def precision_recall_at_k(predictions, k=10, threshold=3.5, verbose=True):
```
verbose parameter is unnecessary

mahermalaeb replied:
I missed this one. Now it should be removed

NicolasHug commented on examples/precision_recall_at_k.py:
```diff
@@ -0,0 +1,73 @@
+                est, true_r) in user_ratings[:k])
+
+        # Count number of recommended items in top k
+        n_rec_k = sum(
```
could this fit in one line?

mahermalaeb replied:
if you meant `n_rec_k` then yes. I have changed it to a one-liner

NicolasHug commented on examples/precision_recall_at_k.py:
```diff
@@ -0,0 +1,73 @@
+        n_rec_k = sum(
+            (est >= threshold) for (est, _) in user_ratings[:k])
+
+        # Precision@K: Proportion of top-k documents that are relevant
```
I'd rather go for "Proportion of recommended items that are relevant"

mahermalaeb replied:
Fixed

NicolasHug commented on examples/precision_recall_at_k.py:
```diff
@@ -0,0 +1,73 @@
+        precision_at_k = n_rel_and_rec_k / n_rec_k if n_rec_k != 0 else 1
+
+        # Recall@K: Proportion of relevant items that are in the top-k
+        recall_at_k = n_rel_and_rec_k / n_rel if n_rel != 0 else 1
```
I'd rather go for "Proportion of relevant items that are recommended"

mahermalaeb replied:
Fixed

NicolasHug commented on examples/precision_recall_at_k.py:
```diff
@@ -0,0 +1,73 @@
+"""
+This module illustrates how to compute Precision at k and Recall at k metrics.
```
I'd rather stop here for the module description and provide more details in the FAQ entry (plus, it says that we "predict all
 +the ratings for the pairs (user, item) that are not in the training set", which is not the case)

mahermalaeb replied:
I kept the first line only

NicolasHug commented on doc/source/FAQ.rst:
```diff
@@ -20,6 +20,23 @@ top-10 prediction for each user.
+a recommendation model. We first split our dataset into train and test 
+folds. Using cross validation and on each iteration we train on our trainset 
+and get the predictions from the testset. Finally we compute Precision@k 
+and Recall@k for each user in predictions.
```
"Finally we compute Precision@k and Recall@k for each user in predictions."

Could be expanded into:

-----------------------------------
Finally, we compute Precision@k and Recall@k for each user:
- Precision@k = #{Recommended items that are relevant} / #{recommended items}
- Recall@k = #{Recommended items that are relevant} / #{relevant items}

An item is considered relevant if its rating is greater than the threshold k. An item is considered recommended if its estimated rating is greater than the threshold k, and if it is among the k highest estimated ratings.

------------------------------
As the file will be proessed by sphinx, we could use latex formatting, e.g.

    precision@k = :math:`\frac{ | \{ \text{Recommended items that are relevant} \} | }{ | \{ \text{Recommended items} \} | }`
Or something like that. I'm not entirely sure about the syntax as I can't check it from github.

mahermalaeb replied:
I have changed the description as you suggested. The latex syntax you gave me is correct!
Note however that I have changed 
> An item is considered relevant if its rating is greater than the threshold k.

to 

> An item is considered relevant if its rating is greater than the user defined threshold. 

That's because the parameter `k` is irrelevant for this definition. You might have put it by mistake. I changed the definition of recommended similarly.

NicolasHug commented on examples/precision_recall_at_k.py:
```diff
@@ -0,0 +1,73 @@
+
+
+data = Dataset.load_builtin('ml-100k')
+
```
remove blank line

mahermalaeb replied:
I have removed the blank line between `data = Dataset.load_builtin('ml-100k')` and `data.split(n_folds=5)`

NicolasHug commented:
Great!
Thanks a lot Maher
